Experimental filter-based rc channel smoothing

etracer65 opened the pull request:
Adds an additional rc channel smoothing algorithm that can be used in place of the default rc interpolation. Utilizing a filter-based approach the smoothing has lower latency and is immune to loop time jitter that can introduce artifacts.  Additionally a smoothing filter is added to the setpoint derivative used to produce D-term setpoint weight resulting in a smoother effect on D.

The default setting is to use the previous interpolation logic and there are no changes unless the optional method is selected.

Configuration:

`rc_smoothing_type`: (INTERPOLATION | FILTER) - defaults to INTERPOLATION

`rc_smoothing_input_hz`: (0-255) - sets the rc channel input filter cutoff in Hz. Default value of 0 will enable auto calculation based on received RX frame rate.

`rc_smoothing_derivative_hz`: (0-255) - sets the setpoint weight derivative filter cutoff in Hz. Default value of 0 will enable auto calculation based on received RX frame rate.

`rc_smoothing_debug_axis`: (ROLL | PITCH | YAW | THROTTLE) - determines which axis is logged in the debug fields

`rc_smoothing_input_type`: (PT1 | BIQUAD) - Determines the type of filter used to filter the input.  Default is BIQUAD.

`rc_smoothing_derivative_type`: (OFF | PT1 | BIQUAD) - Determines the type of filter used to filter the setpoint weight derivative. Default is OFF.

Debug logging:

`set debug_mode = RC_SMOOTHING`

debug(0) = raw un-smoothed rc channel data
debug(1) = raw un-smoothed setpoint derivative
debug(2) = filtered setpoint derivative before applied to setpoint weight
debug(3) = auto-calculated filter cutoff frequency base after sampling the rx frame rate

Notes:

**Currently only enabled for F4/F7 due to flash size limitations**

Uses the `rc_inter_ch` parameter to determine which channels are smoothed (same as the interpolation logic)

The auto filter cutoff calculation will set a cutoff frequency of 30Hz for typical SBUS frames (9ms).  11ms Spektrum will calculate to approximately 25Hz. The user can manually enter the filter cutoffs to be used instead of the auto calculation.  The current default calculation was chosen as a good starting point but may be adjusted in the future.

Setting a lower cutoff frequency will result in more smoothing, but also more delay. Higher cutoffs are the reverse.

There currently isn't any support for receivers that change their rx frame rate dynamically.  So for CRSF users wishing to use this alternate smoothing method should change their settings to lock the rx frame rate for now. Support for auto-adjusting to new frame rates in flight will likely be added.

Adds a framework to allow other types of rc smoothing/interpolation methods to be added in the future.

It may be worthwhile to experiment with different types of filter calculations to see if the results can be improved further.

etracer65 commented:
In the initial release I have removed the PID loop under-sampling that theoretically would allow the filters to be fed a fraction of the PID loop rate - possibly reducing processing time in the PID loop.  The logic added a lot of complexity and it's unclear how much processing time is saved.  This may be added back if warranted.

etracer65 commented:
Great. F7's are overflowing ITCM_RAM. :(

Suggestions?

etracer65 commented:
Some example of the differences between classic interpolation and the filter method.  Top graph is the motors, middle is the roll rcCommand and bottom is debugging showing the setpoint derivative used to derive the setpoint weight applied to D.

First screenshot is interpolation:

![interpolation](https://user-images.githubusercontent.com/17088539/40814489-1fdde566-6530-11e8-9b56-fa07109695ce.png)

Note the trapezoidal shapes in the setpoint derivative and how their flowing through to the motor traces as sharp edges.  They are caused by jitter in the PID loop causing the interpolation steps to be miscalculated leading to slope changes in the rcCommand. Each slope change causes a discontinuity which results in a spike of valley in the derivative.

Next is filter:

![filter](https://user-images.githubusercontent.com/17088539/40814569-74c1cd2c-6530-11e8-9fa9-15e5170746e1.png)

Jitter is no longer a problem because the filter is continuously fed with the most recent data allowing the output to follow the stepped input.  We possibly have some opportunity to improve this even further with different filter types.  PT1 is probably not ideal and the convex/concave sections between each step are the result of PT1 being a non-linear filter.

digitalentity commented:
I've also tried BiQuad filter on RC command and I'm quite happy with the results.

DieHertz commented:
I like biquad better as it might result in continuity of first derivative

digitalentity commented:
@DieHertz indeed, it by itself eliminates dterm spikes. The only problem is that we don't know RC update rate and some receivers are known to change that on the fly, so I had to make code to constantly monitor RC update rate and if it changes - reinitialize the filter.

mikeller commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+}
+
+#ifdef USE_RC_SMOOTHING_FILTER
+FAST_CODE FAST_CODE_NOINLINE uint8_t processRcSmoothingFilter(void)
```
Why is this `FAST_CODE_NOINLINE`? This will just make us slower and cost us more ITCM_RAM (worst case).

DieHertz replied:
Looks like the attributes were copy-pasted.

mikeller commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+
+    uint8_t updatedChannel = 0;
+
+    static float lastRxData[4];
```
Question for @DieHertz: Will static variables in `FAST_CODE` go into `FAST_RAM` automatically, or do they have to be attributed for this?

DieHertz replied:
They won't, and there's absolutely no relationship between the two.
And you may actually want them to go to `SRAM` sometimes rather than `DTCM-RAM`.

mikeller replied:
So declaring `FAST_CODE` will work for static local variables, and will put them into fast RAM?

DieHertz replied:
Not sure I understand your last comment. `FAST_CODE` only affects instructions. Static variables will still go into SRAM unless marked `FASTRAM`

mikeller replied:
Sorry, I meant to say `FAST_RAM` above.

mikeller commented on src/main/flight/pid.c:
```diff
@@ -821,16 +837,29 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+
+#ifdef USE_RC_SMOOTHING_FILTER
+            if (axis == rcSmoothingDebugAxis) {
+                DEBUG_SET(DEBUG_RC_SMOOTHING, 1, lrintf(pidSetpointDelta * 100.0f));
```
All these `DEBUG_SET` in `FAST_CODE` makes me think we should probably make `debug` `FAST_RAM`...

DieHertz commented:
I didn't notice you've extracted `subTaskRcCommand` as `NOINLINE`.
Having it in ordinary `FLASH` instead of `ITCM_RAM` kind of defeats the purpose of optimizations, especially as you are calling `FAST_CODE` functions inside it (`processRcCommand`).

DieHertz commented:
>Great. F7's are overflowing ITCM_RAM. :(
You have added more code to it, what did you expect? :-D

DieHertz commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+                rxFrameTimeSum += currentRxRefreshRate;
+                rxFrameCount++;
+                if (rxFrameCount >= RC_SMOOTHING_FILTER_TRAINING_SAMPLES) {
+                    const float avgRxFrameRate = rxFrameTimeSum / rxFrameCount / 1000.0f;
```
`/ 1000`

DieHertz commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+        }
+        // If the filter cutoffs are set to auto and we have good rx data, then determine the average rx frame rate
+        // and use that to calculate the filter cutoff frequencies
+        if ((filterCutoffFrequency == 0) || (derivativeCutoffFrequency == 0)) {
```
It would make it a bit cleaner to either extract these into `boolean` variables and reuse them here and below in ternary operators, or just use `!value` as condition.

DieHertz commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+
+    // Once we've determined the filter cutoff frequencies then initialize the filters
+    if (!filterInitialized && (targetPidLooptime > 0) && (filterCutoffFrequency != 0) && (derivativeCutoffFrequency != 0)) {
+        const float dT = targetPidLooptime * 0.000001f;
```
I think I asked before to rewrite the float as `1e-6f`

DieHertz commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+
+    for (updatedChannel = ROLL; updatedChannel < interpolationChannels; updatedChannel++) {
+        if (filterInitialized) {
+                rcCommand[updatedChannel] = pt1FilterApply(&rcCommandFilter[updatedChannel], lastRxData[updatedChannel]);
```
I like this more as a ternary operator.
You seem to have indentation off here.

ledvinap replied:
It may be even better to init filter to dummy values (alpha = 0) in `initialized == false` and avoid the test here ...

etracer65 replied:
`filterCutoffFrequency` and `derivativeCutoffFrequency` are not really dummy values.  They get assigned in the first initialization stage from the configuration parameters.
```
    if (!initialized) {
        initialized = true;
        interpolationChannels = rxConfig()->rcInterpolationChannels + 2; //"RP", "RPY", "RPYT"
        filterCutoffFrequency = rxConfig()->rc_smoothing_input_cutoff;
        derivativeCutoffFrequency = rxConfig()->rc_smoothing_derivative_cutoff;
    }
```
The "training" phase needs to run inside of `if (isRXDataNew) {` because we only want to sum when we receive new rx frames.  Once we reach the required number of samples we compute average and only override the configuration cutoff if it's 0 (0 = AUTO).

I'll rearrange the logic to wrap the entire sum, average and filter initialization inside of `if (!filterInitialized) {`

ledvinap commented on src/main/fc/fc_rc.c:
```diff
@@ -239,6 +237,105 @@ FAST_CODE FAST_CODE_NOINLINE void processRcCommand(void)
+    static uint16_t filterCutoffFrequency;
+    static uint16_t derivativeCutoffFrequency;
+    static uint8_t interpolationChannels;
+
```
All initialization tests can run only when new sample is available (saving some tests in pid loop)

etracer65 commented:
@DieHertz 

> I didn't notice you've extracted subTaskRcCommand as NOINLINE.
> Having it in ordinary FLASH instead of ITCM_RAM kind of defeats the purpose of optimizations, especially as you are calling FAST_CODE functions inside it (processRcCommand).

I didn't really do this... well not exactly.  The contents of `subTaskRcCommand()` used to be completely contained inside of `subTaskMainSubprocesses()` and it's defined as `static FAST_CODE_NOINLINE` (not my doing).  When I split out the functions into `subTaskRcCommand()` to allow running the interpolation in a different order, initially I didn't have `FAST_CODE_NOINLINE` and as a result it was inheriting `FAST_CODE` `taskMainPidLoop()` and this caused ITCM_RAM overflows.  So I changed `subTaskRcCommand()` to `FAST_CODE_NOINLINE` to match `subTaskMainSubprocesses()` which the code was previously part of.  So if there's a problem with this then it's really that `subTaskMainSubprocesses()` shouoldn't have been `FAST_CODE_NOINLINE`.

These changes were in #5905 and #5992.

DieHertz commented:
`subTaskMainSubprocesses` shouldn't have been marked `FAST_CODE` as it has a lot of non-critical functionality, but extracting a new interpolation method it should have gone into ITCM.

DieHertz commented:
None below `updateRcCommands` should be `NOINLINE`

etracer65 commented:
So for clarity, is this correct then?
```
FAST_CODE void taskMainPidLoop()
    static FAST_CODE_NOINLINE subTaskMainSubprocesses() 
    static FAST_CODE_NOINLINE subTaskRcCommand()
         FAST_CODE updateRcCommands()
            processRcInterpolation()
            processRcSmoothingFilter()
```
Or do I need to explicitly determine the things below `updateRcCommands()`?  I thought they would inherit so not needed.  `updateRcCommands()` doesn't really do much now except act as a dispatcher for the two smoothing methods.  Previously the interpolation code was contained inside it.

DieHertz commented:
They should be `FAST_CODE` but not `NOINLINE`, for clarity.

etracer65 commented:
@digitalentity @DieHertz I'll experiment with biquad as well as a comparison.  It seemed like a simple moving average might be a good fit but it required a lot of samples to cross rx frames at high pid loop speeds.  Needing at least 72 samples for SBUS (9ms) at 8KHz, and even worse for 16/32K.

etracer65 commented:
@digitalentity I planed to also implement the monitoring of rx frame rate and adjust the filter (probably just change k and let it smoothly converge rather than a fully initializing the state).  Can you point me to the code you used?

The part that is going to be a nightmare are protocols that have a longer frame rate then the minimum desired period of 20ms (50hz) - like PPM (27ms) and DSM 1024 (22ms).  Because in the current code the interpolation gets called at 20ms and then again 2ms later - and it's a mess.

For CRSF that varies between 150hz and 50hz (I think) it should be less of a problem.

ledvinap commented:
@etracer65 : Current RC interpolation is moving average (boxcar) FIR filter

etracer65 commented:
@ledvinap Well kind of.  The problem with the current implementation is that with each new real sample received from the rx, it forecasts a number of steps based on the previous rx frame interval (at least in auto).  Then these steps are used to linearly update the state for each subsequent PID loop until the next rx frame is received and then the process repeats.

The problem is that because of jitter in the PID loop and to a lesser extent the rx frame rate the forecasted number of steps is often incorrect.  The rx data frame is asynchronous with the PID loop (at least for serial protocols).  It's common to have a longer period (becuase of other load) followed by a shorter period (catching up).  When the period is too long and the steps run out, the state is no longer updated (goes flat) to prevent overshoot.  When the period is short, then the steps don't complete and the next calculation for the next series of steps changes the slope.  Every slope change caused by this jitter introduces noise in the derivative used for setpoint weight (spikes and valleys) and are responsible for the trapezoidal shapes I noted above.

The primary change in the filter logic is that we no longer have a rigid forecasted time for the next rx frame to be applied.  We continue to feed the filter with the last actual received value until we get a new one.  This lets it converge more smoothly.

Doing the same but with a simple moving average instead of PT1 works as well, but at the cost of having to maintain lots of samples (across up to 4 axis).

etracer65 commented:
Some examples from earlier research when I was looking at this issue.

Here's a detail showing the slop changes caused by jitter on the interpolation logic and the result on setpoint derivative.

![slope-change](https://user-images.githubusercontent.com/17088539/40847180-2f842388-65ab-11e8-85c7-970d8850d025.jpg)

This next one shows the forecasted steps (bottom graph) changing due to the pid loop jitter.

![interp_step_count jitter](https://user-images.githubusercontent.com/17088539/40847290-816d1a2e-65ab-11e8-9821-0db60bc2b91b.png)

ledvinap commented:
@etracer: Using moving average with less samples than update rate will result in 'flat' part / using more samples will not read final value. Current smoothing is just different representation of identical filter.

As for derivative - passing derivative separated from value may work best - just use difference between last received RC updates as derivative, or 0 if sample is too late.

ledvinap commented:
@etracer65 : Actually I'm wrong in long-estimate case -  current interpolation will spread remaining distance over next step, averaging will apply in overlapping interval (part of estimate overlapping next sample interval).

etracer65 commented:
Now compiles for all F4/F7 targets.  Even compiles for OMNIBUS with 166 bytes to spare, but I'm fine with leaving it F4/F7 for now.  If we have free space then we might consider enabling F3 for one of the RC's.

etracer65 commented:
Testing with BIQUAD looks very promising.  Added options to select the filter type for both the input and derivative filters.

`rc_smoothing_input_type = PT1 | BIQUAD` (default is BIQUAD)
`rc_smoothing_derivative_type = OFF | PT1 | BIQUAD` (default is OFF)

Set the default to use BIQUAD for the input and turn OFF the derivative filter.  At the auto-calculated cutoff frequency this seems to work well.  Using higher cutoff frequencies for the input may necessitate enabling the derivative filter.

etracer65 commented:
Goal is to have these options for testing/evaluation but to decide on a single working solution and simplify the configuration parameters.

DieHertz commented:
Need to take F4 FAST_RAM usage into account as well, but in general I'd say use it by default inside critical functions.
It's hard to prove it disprove the performance difference when comparing just one variable in SRAM va DTCM-RAM

etracer65 commented:
I will change the static variables in the rc interpolation/rc smoothing functions to use FAST_RAM.  However note that I did a little test in just one function and it does increase flash usage a little.  Probably not a problem, but something to be aware of.

Seeing the effect on F4 (target OMNIBUSF4SD): The before state:
```Memory region         Used Size  Region Size  %age Used
           FLASH:       12453 B        16 KB     76.01%
    FLASH_CONFIG:          0 GB        16 KB      0.00%
          FLASH1:      336324 B       992 KB     33.11%
   SYSTEM_MEMORY:          0 GB        29 KB      0.00%
             RAM:       69576 B       128 KB     53.08%
             CCM:        6528 B        64 KB      9.96%
     BACKUP_SRAM:          0 GB         4 KB      0.00%
       MEMORY_B1:          0 GB         0 GB       nan%
   text	   data	    bss	    dec	    hex	filename
 338493	  10284	  65820	 414597	  65385	./obj/main/betaflight_OMNIBUSF4SD.elf
```

Changing fc/fc_rc.c processRcSmoothingFilter():
```
    static FAST_RAM_ZERO_INIT float lastRxData[4];
    static FAST_RAM_ZERO_INIT pt1Filter_t rcCommandFilterPt1[4];
    static FAST_RAM_ZERO_INIT biquadFilter_t rcCommandFilterBiquad[4];
    static FAST_RAM bool initialized = false;
    static FAST_RAM bool filterInitialized = false;
    static FAST_RAM_ZERO_INIT float rxFrameTimeSum;
    static FAST_RAM_ZERO_INIT int rxFrameCount;
    static FAST_RAM_ZERO_INIT uint16_t defaultCutoffFrequency;
    static FAST_RAM_ZERO_INIT uint16_t filterCutoffFrequency;
    static FAST_RAM_ZERO_INIT uint16_t derivativeCutoffFrequency;
    static FAST_RAM_ZERO_INIT uint8_t interpolationChannels;
```

After result uses an extra 80 bytes of FLASH:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       12453 B        16 KB     76.01%
    FLASH_CONFIG:          0 GB        16 KB      0.00%
          FLASH1:      336404 B       992 KB     33.12%
   SYSTEM_MEMORY:          0 GB        29 KB      0.00%
             RAM:       69360 B       128 KB     52.92%
             CCM:        6752 B        64 KB     10.30%
     BACKUP_SRAM:          0 GB         4 KB      0.00%
       MEMORY_B1:          0 GB         0 GB       nan%
   text	   data	    bss	    dec	    hex	filename
 338565	  10292	  65820	 414677	  653d5	./obj/main/betaflight_OMNIBUSF4SD.elf
```

DieHertz commented:
I'd like to know where the FLASH usage increase comes from. Could you find out?

etracer65 commented:
@DieHertz 

Did a controlled test and simply changing the statics in fc/fc_rc.c processRcSmoothingFilter() to `FAST_RAM_ZERO_INIT` results in this change:
```
< taskMainPidLoop 0x00005668
---
> taskMainPidLoop 0x0000569c
```

DieHertz commented:
Thank you, I'll add assembly dumps into script as well, it will at least help collectimg diagnostics and e.g. passing them to me.

mikeller commented:
Fixes #4233.

joelucid commented:
I've been staring at these graphs for a long time trying to find out how a linear interpolation of the setpoint rate between two rc commands could ever lead to a trapezoidal first derivative. Doesn't make sense, right? Derivative of a linearly interpolated function should be a constant between two points. It should look like a series of steps, having zero slope between interpolation points.

Also I did a prototype of extrapolating rc and doing the setpoint weight interpolation in the pid controller some time ago (https://github.com/joelucid/betaflight/tree/rc_extra). That looked exactly like the expected steps.

Just now I realized what's going on: the rc interpolation code interpolates rc command, not setpoint. Since setpoint is derived from rc exponentially via srates that leads to a sloped curve between interpolation points.

I suspect that at least for d term weight it's preferable to interpolate setpoint rate. The reason is that small variances in slope in rc command (caused by jitter etc) will otherwise result in large variances in d term weight due to the magnifying effect of srates.

Here's a pic of a roll from my rc_extra prototype:

![image uploaded from ios](https://user-images.githubusercontent.com/18134492/40886429-eef67490-6737-11e8-8c05-b59975300ab9.jpg)

This also uses smart_feedforward so if P and dterm weight have the same sign the larger is taken. You can see how dterm weight is a clean sequence of steps, only showing some slope due to the gyro D component.

In general I'm not a big fan of pt1 filtering d weight since it will result in attenuation of the integrated d weight between two setpoints depending on the speed in which the stick was moved from a to b. That is physically wrong, so you will likely find that in order to avoid overshoots for slow moves you'll be forced to reduce d weight so much that you don't have enough for quick moves.

I think it would be better to spread out the d weight you get from the step change in rc input over a number of loops such that the total amount stays the same. See pid.c in rc_extra for how I've done that - basically the same as what rc interpolation now does on the rc command, but on setpoint.

Then there will still be the issue of what to do when the rx misses a packet, causing d term weight to drop down to zero. I'm less sure about what to do in that case. With frsky and 9ms rx packet distance I'd say it would likely be best to extrapolate rx just as I've done between packets in rc_extra.

joelucid commented:
Also interpolating setpoint rate results in a more front-loaded output than interpolating rc command. So the current interpolation approach effectively adds latency to the control response. It does not make sense to apply the rate scaling from our TX also to how we apply feed forward. If anything we would want to front-load how we apply it to shorten response times.

etracer65 commented on src/main/pg/rx.h:
```diff
@@ -49,7 +49,7 @@ typedef struct rxConfig_s {
     uint8_t max_aux_channel;
     uint8_t rssi_src_frame_errors;          // true to use frame drop flags in the rx protocol
-
+    int8_t rssi_offset;                     // offset applied to the RSSI value before it is returned
```
I don't know what's going on.  Can't get the rebase to work and it seems to be pulling code from an older commit.  That's a mistake from when I was trying to fix it and get it to merge properly.  The conflicts in github look correct but the rebase locally is missing code.

For example: Github conflicts for pg/rx.h (correct)
```
    uint16_t rx_min_usec;
    uint16_t rx_max_usec;
    uint8_t max_aux_channel;
    uint8_t rssi_src_frame_errors;          // true to use frame drop flags in the rx protocol
    int8_t rssi_offset;                     // offset applied to the RSSI value before it is returned
<<<<<<< rc_smoothing_filter
    uint8_t rc_smoothing_type;              // Determines the smoothing algorithm to use: INTERPOLATION or FILTER
    uint8_t rc_smoothing_input_cutoff;      // Filter cutoff frequency for the input filter (0 = auto)
    uint8_t rc_smoothing_derivative_cutoff; // Filter cutoff frequency for the setpoint weight derivative filter (0 = auto)
    uint8_t rc_smoothing_debug_axis;        // Axis to log as debug values when debug_mode = RC_SMOOTHING
    uint8_t rc_smoothing_input_type;        // Input filter type (0 = PT1, 1 = BIQUAD)
    uint8_t rc_smoothing_derivative_type;   // Derivative filter type (0 = OFF, 1 = PT1, 2 = BIQUAD)
=======
>>>>>>> master
```

From my local rebase (incorrect):
```
    uint16_t rx_min_usec;
    uint16_t rx_max_usec;
    uint8_t max_aux_channel;
    uint8_t rssi_src_frame_errors;          // true to use frame drop flags in the rx protocol
<<<<<<< HEAD
    int8_t rssi_offset;                     // offset applied to the RSSI value before it is returned
=======

    uint8_t rc_smoothing_type;              // Determines the smoothing algorithm to use: INTERPOLATION or FILTER
    uint8_t rc_smoothing_input_cutoff;      // Filter cutoff frequency for the input filter (0 = auto)
    uint8_t rc_smoothing_derivative_cutoff; // Filter cutoff frequency for the setpoint weight derivative filter (0 = auto)
    uint8_t rc_smoothing_debug_axis;        // Axis to log as debug values when debug_mode = RC_SMOOTHING
>>>>>>> f9c7429... Experimental filter-based rc channel smoothing
```

My local repo is up to date with remote and the PR matches so I'm not sure what's going on - never seen this before.

jcutcher01 commented:
My three every-day quads are running crossfire and I wasn't aware that the protocol is dynamically changing the frame rate.  Can someone elaborate on this? The 3.4 Tuning tips doesn't mention crossfire.

etracer65 commented:
@jcutcher01 Under low signal conditions Crossfire can drop to a lower rx frame rate.  Normally it's 150Hz but can drop to 50Hz under these conditions (it's either or - not a scaled range).  Currently RC Smoothing won't handle this change so it's recommended that you don't use it if you're planning to use Crossfire in long-range situations.  Support for the changing frame rate will be added.

jcutcher01 commented:
Thank You @etracer65 I understand now what the issue is. I look forward to the new code when it’s implemented.
